File: idealingua/idealingua-runtime-rpc-typescript/src/main/resources/runtime/typescript/irt/formatter.ts

import * as moment from 'moment';

export class Formatter {
    public static readonly DATETIME_FORMATS = [
        'YYYY-MM-DDTHH:mm:ssZ',
        'YYYY-MM-DDTHH:mm:ss.SZ',
        'YYYY-MM-DDTHH:mm:ss.SSZ',
        'YYYY-MM-DDTHH:mm:ss.SSSZ',
        'YYYY-MM-DDTHH:mm:ss.SSSSZ',
        'YYYY-MM-DDTHH:mm:ss.SSSSSZ',
        'YYYY-MM-DDTHH:mm:ss.SSSSSSZ',
        'YYYY-MM-DDTHH:mm:ss.SSSSSSSZ',
        'YYYY-MM-DDTHH:mm:ss.SSSSSSSSZ',
        'YYYY-MM-DDTHH:mm:ss.SSSSSSSSSZ',
        'YYYY-MM-DDTHH:mm:ss',
        'YYYY-MM-DDTHH:mm:ss.S',
        'YYYY-MM-DDTHH:mm:ss.SS',
        'YYYY-MM-DDTHH:mm:ss.SSS',
        'YYYY-MM-DDTHH:mm:ss.SSSS',
        'YYYY-MM-DDTHH:mm:ss.SSSSS',
        'YYYY-MM-DDTHH:mm:ss.SSSSSS',
        'YYYY-MM-DDTHH:mm:ss.SSSSSSS',
        'YYYY-MM-DDTHH:mm:ss.SSSSSSSS',
        'YYYY-MM-DDTHH:mm:ss.SSSSSSSSS',
    ];

    public static readTime(value: string): Date {
        return moment(value, 'HH:mm:ss.SSS').toDate();
    }

    public static writeTime(value: Date): string {
        return moment(value).format('HH:mm:ss.SSS');
    }

    public static readDate(value: string): Date {
        return moment(value, 'YYYY-MM-DD').toDate();
    }

    public static writeDate(value: Date): string {
        return moment(value).format('YYYY-MM-DD');
    }

    public static readDateTime(value: string, utc: boolean = false): Date {
        const regionIndex = value.indexOf('[');
        if (regionIndex >= 0) {
            // For the time being, we just ignore [Europe/Dublin] kind of regions
            value = value.substring(0, regionIndex);
        }

        const res = moment(value, Formatter.DATETIME_FORMATS);
        return utc ? res.utc().toDate() : res.toDate();
    }

    public static readZoneDateTime(value: string): Date {
        return Formatter.readDateTime(value);
    }

    public static writeZoneDateTime(value: Date): string {
        return moment(value).format('YYYY-MM-DDTHH:mm:ss.SSSZ');
    }

    public static readLocalDateTime(value: string): Date {
        return Formatter.readDateTime(value);
    }

    public static writeLocalDateTime(value: Date): string {
        return moment(value).format('YYYY-MM-DDTHH:mm:ss.SSS');
    }

    public static readUTCDateTime(value: string): Date {
        return Formatter.readDateTime(value, true);
    }

    public static writeUTCDateTime(value: Date): string {
        return moment(value).utc().format('YYYY-MM-DDTHH:mm:ss.SSSZ');
    }
}

File: idealingua-v1/idealingua-v1-runtime-rpc-go/src/main/resources/runtime/go/irt/formatter.go
package irt

import (
	"strings"
	"time"
)

// TODO Replace with better logic than brute-forcing this thing
var DATETIME_FORMATS = []string{
	"2006-01-02T15:04:05Z07:00",
	"2006-01-02T15:04:05.0Z07:00",
	"2006-01-02T15:04:05.00Z07:00",
	"2006-01-02T15:04:05.000Z07:00",
	"2006-01-02T15:04:05.0000Z07:00",
	"2006-01-02T15:04:05.00000Z07:00",
	"2006-01-02T15:04:05.000000Z07:00",
	"2006-01-02T15:04:05.0000000Z07:00",
	"2006-01-02T15:04:05.00000000Z07:00",
	"2006-01-02T15:04:05.000000000Z07:00",
	"2006-01-02T15:04:05Z",
	"2006-01-02T15:04:05.0Z",
	"2006-01-02T15:04:05.00Z",
	"2006-01-02T15:04:05.000Z",
	"2006-01-02T15:04:05.0000Z",
	"2006-01-02T15:04:05.00000Z",
	"2006-01-02T15:04:05.000000Z",
	"2006-01-02T15:04:05.0000000Z",
	"2006-01-02T15:04:05.00000000Z",
	"2006-01-02T15:04:05.000000000Z",
	"2006-01-02T15:04:05",
	"2006-01-02T15:04:05.0",
	"2006-01-02T15:04:05.00",
	"2006-01-02T15:04:05.000",
	"2006-01-02T15:04:05.0000",
	"2006-01-02T15:04:05.00000",
	"2006-01-02T15:04:05.000000",
	"2006-01-02T15:04:05.0000000",
	"2006-01-02T15:04:05.00000000",
	"2006-01-02T15:04:05.000000000",
}

func ReadTime(value string) (time.Time, error) {
	return time.Parse("15:04:05.000", value)
}

func WriteTime(value time.Time) string {
	return value.Format("15:04:05.000")
}

func ReadDate(value string) (time.Time, error) {
	return time.Parse("2006-01-02", value)
}

func WriteDate(value time.Time) string {
	return value.Format("2006-01-02")
}

func ReadDateTime(value string, utc bool) (time.Time, error) {
	regionIndex := strings.Index(value, "[")
	var region *time.Location
	var err error

	if regionIndex >= 0 {
		region, err = time.LoadLocation(value[regionIndex : len(value)-1])
		if err != nil {
			region = nil
		}
		value = value[:regionIndex]
	}

	var t time.Time

	if region != nil {
		for _, f := range DATETIME_FORMATS {
			t, err = time.ParseInLocation(f, value, region)
			if err == nil {
				return t, nil
			}
		}
	} else {
		for _, f := range DATETIME_FORMATS {
			t, err = time.Parse(f, value)
			if err == nil {
				return t, nil
			}
		}
	}

	if err != nil {
		return t, err
	}

	if utc {
		return t.UTC(), nil
	}

	return t, nil
}

func ReadLocalDateTime(value string) (time.Time, error) {
	return ReadDateTime(value, false)
}

func WriteLocalDateTime(value time.Time) string {
	return value.Format("2006-01-02T15:04:05.000")
}

func ReadZoneDateTime(value string) (time.Time, error) {
	return ReadDateTime(value, false)
}

func WriteZoneDateTime(value time.Time) string {
	return value.Format("2006-01-02T15:04:05.000-07:00")
}

func ReadUTCDateTime(value string) (time.Time, error) {
	return ReadDateTime(value, true)
}

func WriteUTCDateTime(value time.Time) string {
	return value.UTC().Format("2006-01-02T15:04:05.000Z")
}
